packages/datadog-plugin-http/src/client.js

```diff
@@ -145,43 +145,54 @@ function patch (http, methodName, tracer, config) {
     })
   }
 
-  function extractUrl (options) {
-    const uri = options
-    const agent = options.agent || http.globalAgent
-
-    return typeof uri === 'string' ? uri : url.format({
-      protocol: options.protocol || agent.protocol,
-      hostname: options.hostname || options.host || 'localhost',
-      port: options.port,
-      pathname: options.path || options.pathname || '/'
-    })
-  }
-
   function normalizeArgs (inputURL, inputOptions, callback) {
-    let options = normalizeURL(inputURL)
+    if (typeof inputURL === 'string') {
+      try {
+        inputURL = urlToOptions(new url.URL(inputURL))
+      } catch (e) {
+        inputURL = url.parse(inputURL)
+      }
+    } else if (inputURL instanceof url.URL) {
+      inputURL = urlToOptions(inputURL)
+    }
 
     if (typeof inputOptions === 'function') {
       callback = inputOptions
+      inputOptions = inputURL || {}
     } else if (typeof inputOptions === 'object') {
-      options = Object.assign(options, inputOptions)
+      inputURL = Object.assign(inputURL || {}, inputOptions)
     }
-    const uri = extractUrl(options)
-    return { uri, options, callback }
-  }
 
-  function normalizeURL (inputURL) {
-    let options
+    // normalize getHeaders
+    inputURL.headers = inputURL.headers || {}
 
-    if (typeof inputURL === 'string') {
-      options = url.parse(inputURL)
-    } else {
-      options = {}
-      for (const key in inputURL) {
-        options[key] = inputURL[key]
-      }
-    }
+    const uri = url.format(inputURL)
+
+    return { uri, options: inputURL, callback }
+  }
 
-    options.headers = options.headers || {}
+  // https://github.com/nodejs/node/blob/7e911d8b03a838e5ac6bb06c5b313533e89673ef/lib/internal/url.js#L1271
+  function urlToOptions (url) {
+    const agent = url.agent || http.globalAgent
```

Comment:
we have some defaults in the now removed `extractURL` method that I tried to add in here to maintain parity, mainly though this is in line with the internal `urlToOptions` helper method `https://github.com/nodejs/node/blob/7e911d8b03a838e5ac6bb06c5b313533e89673ef/lib/internal/url.js#L1271`